feat(category_theory/types): equiv_equiv_iso (merge #1438 first)

kim-em opened the pull request:
Just tying up a small loose end.

kim-em committed:
trying to use bundled homs

kim-em committed:
bundled monoids should use bundled homs

kim-em committed:
fixes

kim-em committed:
Merge branch 'master' into Mon_bundled_hom

kim-em committed:
fixes

urkud committed:
refactor(*): rewrite concrete categories

urkud committed:
Add module documentation

kim-em committed:
local instance

urkud committed:
Merge branch 'Mon_bundled_hom' into concrete-categories

Compile fails. I'll resolve actual conflicts later.

urkud committed:
Move files around; don't touch content yet

urkud committed:
Move code around, fix some compile errors

urkud committed:
Add `unbundled_hom.lean`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Turn `hom` into an argument of `(un)bundled_hom`

For some reason, it helps Lean find coercions from `Ring` category
morphisms to functions.

urkud committed:
Define `unbundled_hom.mk_has_forget`; fix compile

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Add some documentation

urkud committed:
Fix compile

kim-em committed:
chore(category_theory): require morphisms live in Type

kim-em committed:
move back to Type

kim-em committed:
tweaks to doc-comments

kim-em committed:
fixing some formatting

urkud committed:
Revert most of the changes to `data/mv_polynomials`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Drop unused argument, add a comment

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
merge

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
Merge remote-tracking branch 'origin/master' into category_no_sorts

urkud committed:
Simplify universe levels in `concrete_category/basic`.

Still fails to compile.

urkud committed:
Simplify universe levels in `concrete_category/{un,}bundled_hom`

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

kim-em committed:
fixes

urkud committed:
Fix an `import`

urkud committed:
Merge branch 'master' into concrete-categories

urkud committed:
Doc cleanup

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Merge branch 'master' into concrete-categories

kim-em committed:
feat(order/category/DecLinOrder): updates to concrete_category, and instances of order categories.

kim-em committed:
feat(category_theory/types): equiv_equiv_iso

kim-em committed:
comments

kim-em committed:
fix

kim-em committed:
fix

kim-em committed:
fix

kim-em committed:
fixes

kim-em committed:
...

kim-em committed:
...

jcommelin commented on src/category_theory/types.lean:
```diff
@@ -170,3 +170,17 @@ def to_equiv (i : X ≅ Y) : X ≃ Y :=
+-- (In this order the proofs are handled by `obviously`.)
+
+/-- equivalences (between types in the same universe) are the same as (isomorphic to) isomorphisms of types -/
+def equiv_iso_iso {X Y : Sort u} : (X ≃ Y) ≅ (X ≅ Y) :=
```
I would expect some simp-lemmas for the projections of the definitions... does it make sense to add those?

kim-em replied:
Added. Also, made the arguments to `equiv_equiv_iso` explicit, because otherwise the coercion to function won't fire (and added a comment to this effect).

jcommelin replied:
They weren't added on my version of Github... did you push?

kim-em replied:
Sorry, hopefully all there now.

Commit:
simp lemmas

Commit:
add Aut_mul_equiv_perm

jcommelin commented on src/category_theory/types.lean:
```diff
@@ -182,5 +183,22 @@ def equiv_iso_iso {X Y : Sort u} : (X ≃ Y) ≅ (X ≅ Y) :=
+
+namespace category_theory
+
+def Aut_mul_equiv_perm {α : Type u} : Aut α ≃* equiv.perm α :=
```
You could do a similar cheat as with `equiv_equiv_iso`...
First define the thing as an iso in `Mon` or `Group` or whatever, and then deduce this one. It'll give you your auto_params (-;

kim-em replied:
I thought about that, but then this file would have to import `algebra.category.Group`, or this content would have to move there, and my head started hurting.

jcommelin replied:
But we'll probably want the groupy version of `equiv_equiv_iso` somewhere at some point...

kim-em replied:
Okay, I'm succumbing to feature creep and adding this now.

kim-em replied:
Okay, all the multiplicative versions are here now as well. There was no point doing this without rebasing on top of @urkud's (lovely) work on concrete categories, so this PR will now have to wait until #1380 and #1438 have bene merged.

kim-em committed:
Merge branch 'category_no_sorts' into equiv_equiv_iso

kim-em committed:
update post #1412

kim-em committed:
Merge branch 'category_no_sorts' into concrete-categories

kim-em committed:
update post #1412

kim-em committed:
Merge remote-tracking branch 'origin/master' into concrete-categories

kim-em committed:
merge

urkud committed:
Drop `simp`

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Rename variable

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Rename variable

Co-Authored-By: Johan Commelin <johan@commelin.net>

Commit:
Merge remote-tracking branch 'urkud/concrete-categories' into DecLinOrder

urkud committed:
Fix more issues reported by @rwbarton

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Rename variable

Co-Authored-By: Reid Barton <rwbarton@gmail.com>

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
Merge branch 'master' into concrete-categories

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

Commit:
merge

Commit:
blech

Commit:
merge

urkud committed:
Use `subtype.eq` instead of `subtype.ext.2`

urkud committed:
Cleanup

urkud committed:
Fix compile: now `ring_hom.ext` takes fewer explicit args

urkud committed:
Fix compile

urkud committed:
Run `sanity_check` on all files modified in this branch

* Fix most of the issues reported

* Except for the old ones in `mv_polynomial` (postponed to a separate
  PR) and a false positive in `single_obj`

Commit:
ring_hom.ext changed signature

Commit:
fix signatures

Commit:
more signatures

Commit:
breaking everything

kim-em committed:
breaking more things

kim-em committed:
merge

kim-em committed:
looks okay

kim-em committed:
the ring versions, too

kim-em committed:
naming

kim-em committed:
fix

urkud committed:
Review `∀` vs `Π`

urkud committed:
Remove some unnecessary parentheses

kim-em committed:
add comment

urkud committed:
punit_instances: no need to explicitly provide constants and operations

urkud committed:
Rename `has_forget` to `has_forget₂`

kim-em committed:
add comment, simplify comm_ring punit

urkud committed:
Merge branch 'concrete-categories' of ssh://github.com/urkud/mathlib into concrete-categories

urkud committed:
Drop `private def free_obj`

kim-em committed:
documentation

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
merge

kim-em committed:
fixes

kim-em committed:
fix typo

kim-em committed:
fix

kim-em committed:
merge

kim-em committed:
partial fixes, probably all still borked

kim-em committed:
ugh, I mangled the merge

kim-em committed:
fixes

kim-em committed:
comments on uniform space categories

kim-em committed:
cleanup

kim-em committed:
fix botched merge

kim-em committed:
Merge branch 'DecLinOrder' into equiv_equiv_iso

kim-em committed:
Update src/algebra/category/Mon/basic.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Update src/order/category/DecLinOrder.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Update src/algebra/category/Mon/basic.lean

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
change examples to `concrete_category`

kim-em committed:
Rename types to match mathematical usage

Co-Authored-By: Johan Commelin <johan@commelin.net>

kim-em committed:
Merge branch 'DecLinOrder' of github.com:leanprover-community/mathlib into DecLinOrder

kim-em committed:
fixing cases

kim-em committed:
rename notation to `U`

kim-em committed:
make module doc string

kim-em committed:
solved coercion problem

kim-em committed:
experiments

kim-em committed:
huh, maybe okay

kim-em committed:
cleanup of UniformSpace

kim-em committed:
minor cleanup

kim-em committed:
Merge branch 'master' into DecLinOrder

kim-em committed:
merge

kim-em committed:
comment

urkud committed:
Add a docstring, reuse an instance

urkud committed:
Formulate naturality of (co)cones over functors to concrete categories without `bundled`

It was here for cocones under the name `naturality_bundled'`. Rename
it `naturality_concrete`, drop the `bundled` version.

urkud committed:
Fix a typo

kim-em committed:
Merge branch 'DecLinOrder' of github.com:leanprover-community/mathlib into DecLinOrder

kim-em committed:
only reducible locally

kim-em committed:
Merge branch 'DecLinOrder' into equiv_equiv_iso

kim-em committed:
add documentation, and use to_additive

kim-em committed:
oops, update Group and Mon too

kim-em committed:
merge and fix doc strings

kim-em committed:
fixes

kim-em committed:
fixes

Commit:
cleanup, rfl lemma

Commit:
Merge branch 'DecLinOrder' into equiv_equiv_iso

kim-em committed:
Merge branch 'master' into DecLinOrder

Commit:
merge
